pkg/flow/flow_components.go

```diff
@@ -0,0 +1,240 @@
+	"github.com/grafana/agent/pkg/flow/internal/controller"
+	"github.com/grafana/agent/pkg/flow/internal/dag"
+	"github.com/grafana/agent/pkg/river/encoding/riverjson"
+)
+
+// ComponentID is a globally unique identifier for a component.
+type ComponentID struct {
+	ModuleID string // Unique ID of the module that the component is running in.
+	LocalID  string // Local ID of the component, unique to the module it is running in.
+}
+
+// String returns the "<ModuleID>/<LocalID>" string representation of the id.
+func (id ComponentID) String() string {
+	if id.ModuleID == "" {
+		return id.LocalID
+	}
+	return id.ModuleID + "/" + id.LocalID
+}
+
+type ComponentDetail struct {
+	// Component is the instance of the component. Component will be nil if the
+	// component exists in the graph but has not successfully evaluated yet.
+	Component component.Component
+
+	ID    ComponentID // ID of the component.
+	Label string      // Component label. Not set for singleton components.
+
+	// References and ReferencedBy are the list of IDs in the same module that
+	// this component depends on, or is depended on by, respectively.
+	References, ReferencedBy []string
+
+	Registration component.Registration // Component registration.
+	Health       component.Health       // Current component health.
+
+	Arguments component.Arguments // Current arguments value of the component.
+	Exports   component.Exports   // Current exports value of the component.
+	DebugInfo interface{}         // Current debug info of the component.
+}
+
+// MarshalJSON returns a JSON representation of cd. The format of the
+// representation is not stable and is subject to change.
+func (cd *ComponentDetail) MarshalJSON() ([]byte, error) {
+	type (
+		componentHealthJSON struct {
+			State       string    `json:"state"`
+			Message     string    `json:"message"`
+			UpdatedTime time.Time `json:"updatedTime"`
+		}
+
+		componentDetailJSON struct {
+			Name         string               `json:"name,omitempty"`
+			Type         string               `json:"type,omitempty"`
+			ID           string               `json:"id,omitempty"`
+			Label        string               `json:"label,omitempty"`
+			References   []string             `json:"referencesTo"`
+			ReferencedBy []string             `json:"referencedBy"`
+			Health       *componentHealthJSON `json:"health"`
+			Original     string               `json:"original"`
+			Arguments    json.RawMessage      `json:"arguments,omitempty"`
+			Exports      json.RawMessage      `json:"exports,omitempty"`
+			DebugInfo    json.RawMessage      `json:"debugInfo,omitempty"`
+		}
+	)
+
+	var (
+		references   = cd.References
+		referencedBy = cd.ReferencedBy
+
+		arguments, exports, debugInfo json.RawMessage
+		err                           error
+	)
+
+	if references == nil {
+		references = []string{}
+	}
+	if referencedBy == nil {
+		referencedBy = []string{}
+	}
+
+	arguments, err = riverjson.MarshalBody(cd.Arguments)
```

Comment:
Sanity check me on these. Looking at the riverjson.MarshalBody tests are we getting the wanted behavior when we don't have arguments, exports, etc. I see the omitempty tags above so maybe that covers it?

<img width="467" alt="image" src="https://github.com/grafana/agent/assets/39704712/3dee4346-19c2-4cf7-ac76-968322f2a37d">

Reply:
I'll wrap them in checks to be safe.